contracts/examples/WitnetRNG.sol

```diff
@@ -0,0 +1,338 @@
+    function cloneDeterministic(bytes32 _salt)
+        public
+        virtual override
+        returns (Clonable _newInstance)
+    {
+        _newInstance = super.cloneDeterministic(_salt);
+        _clone(_newInstance);
+    }
+
+
+    // ================================================================================================================
+    // --- 'Initializable' overriden functions ------------------------------------------------------------------------
+
+    /// @dev Initializes contract's storage context.
+    function initialize(bytes memory _initData)
+        public
+        virtual override
+    {
+        require(address(witnetRandomnessRequest) == address(0), "WitnetRNG: already initialized");
+        witnetRandomnessRequest = WitnetRequestRandomness(
+            abi.decode(
+                _initData,
+                (address)
+            )
+        );
+    }
+
+
+    // ================================================================================================================
+    // --- INTERNAL FUNCTIONS -----------------------------------------------------------------------------------------
+
+    /// @dev Common steps for both deterministic and non-deterministic cloning.
+    function _clone(Clonable _instance) internal {
+        address _request = address(witnetRandomnessRequest.clone());
+        Ownable(_request).transferOwnership(msg.sender);
+        _instance.initialize(abi.encode(_request));
+    }
+
+    /// @dev Returns index of the Most Significant Bit of the given number, applying De Bruijn O(1) algorithm.
+    function _msbDeBruijn32(uint32 _v)
+        internal pure
+        returns (uint8)
+    {
+        uint8[32] memory _bitPosition = [
+                0, 9, 1, 10, 13, 21, 2, 29,
+                11, 14, 16, 18, 22, 25, 3, 30,
+                8, 12, 20, 28, 15, 17, 24, 7,
+                19, 27, 23, 6, 26, 5, 4, 31
+            ];
+        _v |= _v >> 1;
+        _v |= _v >> 2;
+        _v |= _v >> 4;
+        _v |= _v >> 8;
+        _v |= _v >> 16;
+        return _bitPosition[
+            uint32(_v * uint256(0x07c4acdd)) >> 27
+        ];
+    }
+
+    /// @dev Recursively searches for first block after the given one in which a Witnet Randomness request was casted.
+    /// @dev Returns 0 if none found.
+    function _searchNextBlock(uint256 _target, uint256 _latest) internal view returns (uint256) {
+        return ((_target >= _latest) 
+            ? __randomize_[_latest].nextBlock
+            : _searchNextBlock(_target, __randomize_[_latest].prevBlock)
+        );
+    }
+
+    /// @dev Recursively searches for first block before the given one in which a Witnet Randomness request was casted.
+    /// @dev Returns 0 if none found.
+
+    function _searchPrevBlock(uint256 _target, uint256 _latest) internal view returns (uint256) {
+        return ((_target > _latest)
+            ? _latest
+            : _searchPrevBlock(_target, __randomize_[_latest].prevBlock)
```

Comment:
Let's hope that solidity optimizes away the recursion :crossed_fingers:

In some cases this kind of "search" functions can be optimized by providing a "hint" argument, and calculating that "hint" off-chain. This can save some gas in the worst case scenarios, but checking that additional argument also may cost some gas. And also the user contract must be modified to support it, which adds complexity, and I think this contract is already complex enough...

Reply:
The recursion call is within internal methods, so user contract would not need to deal with them at all. Concerning recursion optimization, proposed algorithm requires reading one word, and only one, per recursion step (what's to say ~200 gu per iteration).